extras/minimize_testcase.cpp

```diff
@@ -0,0 +1,344 @@
+  auto result = Triangulate(polys);
+  PolygonParams() = oldParams;
+  for (auto &tri : result) {
+    auto xid = findIndex(polys, tri.x);
+    auto yid = findIndex(polys, tri.y);
+    auto zid = findIndex(polys, tri.z);
+    auto x = polys[xid.first][xid.second];
+    auto y = polys[yid.first][yid.second];
+    auto z = polys[zid.first][zid.second];
+    printf("show(array([[%.7f, %.7f], [%.7f, %.7f], [%.7f, %.7f]]))\n", x.x,
+           x.y, y.x, y.y, z.x, z.y);
+    if (CCW(x, y, z, precision) == -1) {
+      printf("^ not CCW\n");
+    }
+  }
+}
+
+// x direction ray
+// idk which direction, but it is a fixed direction...
+// https://stackoverflow.com/questions/11716268/point-in-polygon-algorithm
+bool rayHit(glm::vec2 point, glm::vec2 q0, glm::vec2 q1) {
+  return ((q0.y >= point.y) != (q1.y >= point.y)) &&
+         (point.x <= (q1.x - q0.x) * (point.y - q0.y) / (q1.y - q0.y) + q0.x);
+}
+
+// get the list of polygons inside polygon i
+// this enumerates over the first point in all other polygons,
+// and check if they are inside the current polygon
+// this algorithm assumes that polygons are non-intersecting
+std::vector<int> getChildren(const Polygons &polys, int i) {
+  std::vector<int> results;
+  for (int j = 0; j < polys.size(); j++) {
+    if (i == j) continue;
+    glm::vec2 point = polys[j][0];
+    auto k1 = [&](size_t k) {
+      return k == (polys[i].size() - 1) ? 0 : (k + 1);
+    };
+    int count = std::count_if(
+#if MANIFOLD_PAR == 'T' && __has_include(<pstl/glue_execution_defs.h>)
+        std::execution::par,
+#endif
+        countAt((size_t)0), countAt(polys[i].size()),
+        [&](size_t k) { return rayHit(point, polys[i][k], polys[i][k1(k)]); });
+    if (count % 2) results.push_back(j);
+  }
+  return results;
+}
+
+// assuming polys is valid
+// we try to remove vertices from polys such that
+// 1. the updated polys is still valid (no overlapping edges, correct winding
+// direction)
+// 2. same error in triangulation (either geometryErr or overlapping triangles)
+void simplify(Polygons &polys, float precision = -1) {
+  bool removedSomething = true;
+  std::string msg;
+  while (removedSomething) {
+    int points = 0;
+    for (auto &poly : polys) points += poly.size();
+    std::cout << "trying to simplify " << points << " points" << std::endl;
+
+    removedSomething = false;
+    // try to remove simple polygons
+    for (int i = 0; i < polys.size(); i++) {
+      std::vector<int> children = getChildren(polys, i);
+      // if there are children, we can't remove it or we will mess up with
+      // winding direction...
+      if (!children.empty()) continue;
+      SimplePolygon poly = std::move(polys[i]);
+      polys.erase(polys.begin() + i);
+      try {
+        auto result = Triangulate(polys, precision);
+        polys.insert(polys.begin() + i, std::move(poly));
+      } catch (geometryErr &e) {
+        if (msg.size() > 0 && msg.compare(e.what()) != 0) {
```

Comment:
What's happening here? The first time we get an error, it proceeds, but if we get it twice then we try something different? Do we need to reset `msg` at some point?

Reply:
We just want to reduce the testcase so we get the same error, this is checking if the error messages are the same.